numpy/core/src/multiarray/sequence.c

```diff
@@ -72,19 +72,26 @@ array_any_nonzero(PyArrayObject *arr)
     npy_intp counter;
     PyArrayIterObject *it;
     npy_bool anyTRUE = NPY_FALSE;
+    int needs_api = 0;
 
     it = (PyArrayIterObject *)PyArray_IterNew((PyObject *)arr);
     if (it == NULL) {
         return anyTRUE;
     }
+    
+    needs_api = PyDataType_FLAGCHK(PyArray_DESCR(arr), NPY_NEEDS_PYAPI);
+
     counter = it->size;
     while (counter--) {
         if (PyArray_DESCR(arr)->f->nonzero(it->dataptr, arr)) {
             anyTRUE = NPY_TRUE;
             break;
         }
+        if (needs_api && PyErr_Occurred()) {
+            break;
+        }
         PyArray_ITER_NEXT(it);
     }
     Py_DECREF(it);
-    return anyTRUE;
+    return PyErr_Occurred() ? -1 : anyTRUE;
```

Comment:
This should be `(needs_api && PyErr_Occured())`, to match the `if`.

I think it'd be clearer to just `DECREF(it); return -1;` instead of the `break` above though